core/arch/aarchxx/mangle.c

```diff
@@ -904,8 +924,54 @@ int
 insert_out_of_line_context_switch(dcontext_t *dcontext, instrlist_t *ilist,
                                   instr_t *instr, bool save)
 {
-    ASSERT_NOT_IMPLEMENTED(false); /* FIXME i#1551, i#1569 */
+#ifdef AARCH64
+    if (save) {
+        /* Reserve stack space to push the context. We do it here instead of
+         * in insert_push_all_registers, so we can save the original value
+         * of X30 on the stack before it is changed by the BL (branch & link)
+         * to the clean call save routine in the code cache.
+         *
+         * sub sp, sp, #clean_call_switch_stack_size
+         */
+        PRE(ilist, instr,
+            XINST_CREATE_sub(dcontext, opnd_create_reg(DR_REG_SP),
+                             OPND_CREATE_INT16(get_clean_call_switch_stack_size())));
+
+        /* stp x30, x30, [sp, #x30_offset] */
+        PRE(ilist, instr,
+            INSTR_CREATE_stp(dcontext,
+                             opnd_create_base_disp(DR_REG_SP, DR_REG_NULL, 0,
+                                                   REG_OFFSET(DR_REG_X30), OPSZ_16),
+                             opnd_create_reg(DR_REG_X30),
+                             opnd_create_reg(DR_REG_X30)));
+    }
+
+    PRE(ilist, instr,
+        INSTR_CREATE_bl
+        (dcontext, save ?
+         opnd_create_pc(get_clean_call_save(dcontext)) :
+         opnd_create_pc(get_clean_call_restore(dcontext))));
```

Comment:
Yes that's true! Unconditional branches use PC relative addressing with a possible range of +-128MB. Do you think it would be sensible to check if the offset between the PC of the previous instruction and the save/restore routines is smaller than 128MB? And if it's bigger, we have to store the target PC in a register and use BR.